discovery_child_development/utils/cluster_analysis_utils.py

```diff
@@ -320,3 +321,116 @@ def cluster_keywords(
         top_cluster_tokens[unique_cluster_labels[i]] = [id_to_token[j] for j in x]
 
     return top_cluster_tokens
+
+
+def get_cluster_centroids(df: pd.DataFrame, embeddings: np.ndarray) -> List[np.ndarray]:
+    """Get the centroids of each cluster
+
+    Args:
+        data_df: Dataframe with integer index [0, 1, 2...] and a column 'cluster' with cluster labels
+        embeddings: Embeddings for each data point
+
+    Returns:
+        List of cluster centroid vectors
+    """
+    centroids = []
+    for i in range(len(df["cluster"].unique())):
+        cluster = df[df["cluster"] == i]
+        centroid = np.mean(embeddings[cluster.index], axis=0)
+        centroids.append(centroid)
+    return centroids
+
+
+def get_n_most_central_vectors(
+    embeddings: np.ndarray, centroid: np.ndarray, n: int = 10
+) -> np.ndarray:
+    """Get the n most similar data points to a centroid"""
+    distances = np.linalg.norm(embeddings - centroid, axis=1)
+    return np.argsort(distances)[:n]
+
+
+def describe_clusters_with_gpt(
+    cluster_df: pd.DataFrame,
+    embeddings: np.ndarray,
+    n_central: int = 15,
+    gpt_message: str = None,
+) -> List[str]:
+    """
+    Generate cluster descriptions from cluster centroids using GPT
+
+    Args:
+        cluster_df (pd.DataFrame): Dataframe with integer index [0, 1, 2...] and columns
+            'cluster' with cluster labels, and 'text' with text data for each data point
+        gpt_message (str): Message to send to GPT to generate cluster description
+        embeddings (np.ndarray): Embeddings for each data point used for determining cluster centroids
+        n_central (int, optional): Number of most central data points to use for cluster description. Defaults to 15.
+
+    Returns:
+        List[str]: List of cluster descriptions
+    """
+    if gpt_message is None:
+        gpt_message = "Here are the most central documents of a document cluster. \
+            Describe what kind of information is this cluster capturing, in 2 sentences. \
+            \n\n##Abstracts\n\n {} \n\n##Description (2 short sentences)"
+
+    # Get cluster centroid indices
+    centroids = get_cluster_centroids(cluster_df, embeddings)
+    most_central = []
+    for i in range(len(centroids)):
+        most_central.append(
+            get_n_most_central_vectors(embeddings, centroids[i], n=n_central)
+        )
+
+    # Generate cluster descriptions
+    cluster_descriptions = []
+    for i in range(len(centroids)):
+        abstracts = cluster_df.iloc[most_central[i]].text.to_list()
+        messages = [
+            {
+                "role": "user",
+                "content": copy.deepcopy(gpt_message).format("\n".join(abstracts)),
```

Comment:
I have heard of deep copy but not super familiar with it in practice! Why is it needed here?

Reply:
I've had issues before when using string formatting, that you format it once and then your string template has been changed.  Here, I want to maintain the template as a template, and create a new string instance every time the template is used.